frontend/src/components/dashboard/hooks/useAutomaticGrouping.ts

```diff
@@ -0,0 +1,322 @@
+  solutions: CurrentAnalysis[];
+}
+
+const meanDistance = (
+  a: IntermediateGroup,
+  b: IntermediateGroup,
+  distanceMap: Map<number, Map<number, number>>,
+): number => {
+  let sum = 0;
+
+  for (const solutionA of a.solutions) {
+    for (const solutionB of b.solutions) {
+      sum +=
+        distanceMap.get(solutionA.solutionId)?.get(solutionB.solutionId) ?? 0;
+    }
+  }
+
+  return sum / (a.solutions.length * b.solutions.length);
+};
+
+export const useAutomaticGrouping = (
+  isAutomaticGrouping: boolean,
+  numberOfGroups: number,
+  solutionsInput: CurrentAnalysis[] | undefined,
+  filters: FilterCriterion[],
+  xAxis: AxesCriterionType,
+  yAxis: AxesCriterionType,
+): (() => Promise<{
+  groupAssignments: SolutionGroupAssignment[];
+  automaticGroups: AutomaticGroup[];
+}>) =>
+  useCallback(async () => {
+    const solutions = solutionsInput?.filter((solution) =>
+      filters.every((filter) => matchesFilter(filter, solution)),
+    );
+
+    if (!solutions || solutions.length === 0 || !isAutomaticGrouping) {
+      return { groupAssignments: [], automaticGroups: [] };
+    }
+
+    let groups: IntermediateGroup[] = solutions.map((solution) => ({
+      solutions: [solution],
+    }));
+
+    const distances: Map<number, Map<number, number>> = new Map();
+    for (const solution of solutions) {
+      distances.set(solution.solutionId, new Map());
+    }
+
+    const promises: Promise<void>[] = [];
+
+    for (let i = 0; i < solutions.length; i++) {
+      for (let j = i + 1; j < solutions.length; j++) {
+        promises.push(
+          computePqDistance(
+            solutions[i].generalAst,
+            solutions[j].generalAst,
+          ).then((distance) => {
+            distances
+              .get(solutions[i].solutionId)!
+              .set(solutions[j].solutionId, distance);
+
+            distances
+              .get(solutions[j].solutionId)!
+              .set(solutions[i].solutionId, distance);
+          }),
+        );
+      }
+    }
+
+    await Promise.all(promises);
+
+    while (groups.length > numberOfGroups) {
+      let minDistance = Number.POSITIVE_INFINITY;
+      const minDistancePair: [number, number] = [-1, -1];
+
+      for (let i = 0; i < groups.length; i++) {
+        for (let j = i + 1; j < groups.length; j++) {
+          const distance = meanDistance(groups[i], groups[j], distances);
+
+          if (distance < minDistance) {
+            minDistance = distance;
+            minDistancePair[0] = i;
+            minDistancePair[1] = j;
+          }
+        }
+      }
+
+      if (minDistancePair[0] < 0 || minDistancePair[1] < 0) {
+        break;
+      }
+
+      const g1 = groups[minDistancePair[0]];
+      const g2 = groups[minDistancePair[1]];
+
+      // merge the two groups with the smallest distance (g1 and g2)
+      groups = [
+        ...groups.filter((g) => g !== g1 && g !== g2),
+        {
+          solutions: [...g1.solutions, ...g2.solutions],
+        },
+      ];
```

Comment:
reading this, I'm wondering if we shouldn't make groups a `Set`, so that we don't have to do the O(n) filtering on every merge, and instead of `minDistancePair: [number, number]` we just keep a reference to the two groups